src/array.jl

```diff
@@ -662,14 +691,22 @@ function Base.resize!(A::CategoricalVector, n::Integer)
     A
 end
 
-function Base.push!(A::CategoricalVector, item)
-    r = get!(A.pool, item)
+function Base.push!(A::CategoricalVector, v::Any)
+    # TODO: use a global table to cache subset relations for all pairs of pools
+    if v isa CategoricalValue && pool(v) !== pool(A) && pool(v) ⊈ pool(A)
+        merge_pools!(A, v)
+    end
+    r = get!(A.pool, v)
     push!(A.refs, r)
     A
 end
 
 function Base.append!(A::CategoricalVector, B::CatArrOrSub)
-    levels!(A, union(levels(A), levels(B)))
+    # TODO: use a global table to cache subset relations for all pairs of pools
+    if pool(B) !== pool(A) && pool(B) ⊈ pool(A)
+        merge_pools!(A, B)
+    end
+    # TODO: optimize recoding
```

Comment:
Is is safe to use use `@inbounds` in line 713 (and worth it? - if not then better leave as is).

Reply:
Safe yes, but as the TODO indicates if we want something fast we should do as in other places and compute a mapping between source and destination refs. Adding `@inbounds` would probably not help as we do a dict lookup for each entry (!).

Maybe calling  `copyto!` would be enough here.